pkg/token/serviceprincipaltoken.go

```diff
@@ -55,28 +61,46 @@ func newServicePrincipalToken(oAuthConfig adal.OAuthConfig, clientID, clientSecr
 	}, nil
 }
 
+// Token fetches an azcore.AccessToken from the Azure SDK and converts it to an adal.Token for use with kubelogin.
 func (p *servicePrincipalToken) Token() (adal.Token, error) {
-	emptyToken := adal.Token{}
-	callback := func(t adal.Token) error {
-		return nil
-	}
+	return p.TokenOptions(nil)
+}
 
-	var (
-		spt *adal.ServicePrincipalToken
-		err error
-	)
+func (p *servicePrincipalToken) TokenOptions(options *azcore.ClientOptions) (adal.Token, error) {
+	emptyToken := adal.Token{}
+	var spnAccessToken azcore.AccessToken
 
+	// Request a new Azure token provider for service principal
 	if p.clientSecret != "" {
-		spt, err = adal.NewServicePrincipalToken(
-			p.oAuthConfig,
+		clientOptions := &azidentity.ClientSecretCredentialOptions{}
+		if options == nil {
+			clientOptions = &azidentity.ClientSecretCredentialOptions{}
+		} else {
+			clientOptions.ClientOptions = *options
+		}
+		cred, err := azidentity.NewClientSecretCredential(
+			p.tenantID,
 			p.clientID,
 			p.clientSecret,
-			p.resourceID,
-			callback)
+			clientOptions,
+		)
+		if err != nil {
+			return emptyToken, fmt.Errorf("unable to create credential. Received: %v", err)
+		}
+
+		// Use the token provider to get a new token
+		spnAccessToken, err = cred.GetToken(context.Background(), policy.TokenRequestOptions{Scopes: []string{p.resourceID + "/.default"}})
 		if err != nil {
-			return emptyToken, fmt.Errorf("failed to create service principal token using secret: %s", err)
+			return emptyToken, fmt.Errorf("expected an empty error but received: %v", err)
 		}
+
 	} else if p.clientCert != "" {
+		clientOptions := &azidentity.ClientCertificateCredentialOptions{}
+		if options == nil {
```

Comment:
besides the same comment earlier, you need to set [`SendCertificateChain`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity#ClientCertificateCredentialOptions) to true to enable SNI auth.

Reply:
Fixed